test/test_mixed_null_assertions.cpp

```diff
@@ -78,3 +78,202 @@ TEST(List_Mixed_do_insert)
     list.insert_null(0);
     list.insert(0, Mixed("hello world"));
 }
+
+TEST(Mixed_List_unresolved_as_null)
+{
+    Group g;
+    auto t = g.add_table("foo");
+    t->add_column_list(type_Mixed, "mixeds");
+    auto obj = t->create_object();
+    auto obj1 = t->create_object();
+
+    auto list = obj.get_list<Mixed>("mixeds");
+
+    list.insert_null(0);
+    list.insert(1, Mixed{"test"});
+    list.insert(2, obj1);
+    obj1.invalidate();
+
+    CHECK(list.size() == 3);
+
+    {
+        // find all mixed nulls or unresolved link should work the same way
+        std::vector<size_t> found;
+        std::vector<size_t> expected = {0, 2};
+        auto check_results = [&]() {
+            CHECK_EQUAL(found.size(), expected.size());
+            std::sort(found.begin(), found.end());
+            std::sort(expected.begin(), expected.end());
+            for (size_t i = 0; i < found.size(); ++i) {
+                CHECK_EQUAL(found[i], expected[i]);
+            }
+        };
+        list.find_all(realm::null(), [&](size_t pos) {
+            found.push_back(pos);
+        });
+        check_results();
+        found = {};
+        list.find_all(obj1, [&](size_t pos) {
+            found.push_back(pos);
+        });
+        check_results();
+    }
+
+    {
+        // find null or find unresolved link diverge, different objects should be returned
+        auto index = list.find_any(realm::null());
+        CHECK(index == 0);
+        index = list.find_first(obj1);
+        CHECK(index == 2);
+        // but both should look like nulls
+        CHECK(list.is_null(0));
+        CHECK(list.is_null(2));
+    }
+
+    {
+        std::vector<size_t> indices{0, 1, 2};
+        list.sort(indices);
+        CHECK(indices.size() == 3);
+        CHECK(indices.at(0) == 0);
+        CHECK(indices.at(1) == 2);
+        CHECK(indices.at(2) == 1);
+        CHECK(list.is_null(indices[0]));
+        CHECK(list.is_null(indices[1]));
+        CHECK(!list.is_null(indices[2]));
+    }
+
+    {
+        std::vector<size_t> indices{0, 1, 2};
+        list.distinct(indices);
+        CHECK(indices.size() == 2);
+        CHECK(indices.at(0) == 0);
+        CHECK(indices.at(1) == 1);
+        CHECK(list.is_null(indices[0]));
+        CHECK(!list.is_null(indices[1]));
+        CHECK(list.find_any(realm::null()) == 0);
+    }
+
+    {
+        list.remove(0);
+        CHECK(list.find_any(obj1) == 1);
+        list.remove(1);
+        CHECK(list.find_any(realm::null()) == npos);
+        CHECK(list.size() == 1);
+    }
+}
+
+TEST(Mixed_Set_unresolved_as_null)
+{
+    Group g;
+
+    auto t = g.add_table("foo");
+    t->add_column_set(type_Mixed, "mixeds");
+    auto obj = t->create_object();
+    auto obj1 = t->create_object();
+    auto obj2 = t->create_object();
+    obj1.invalidate();
```

Comment:
invalidate the object after inserting it into the set, otherwise you just insert an actual null value